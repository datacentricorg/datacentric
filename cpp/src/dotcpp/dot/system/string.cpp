/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <dot/precompiled.hpp>
#define FMT_HEADER_ONLY
#include <fmt/format.h>
#include <dot/implement.hpp>
#include <dot/system/string.hpp>
#include <dot/system/object.hpp>
#include <dot/system/nullable.hpp>
#include <dot/system/type.hpp>

namespace dot
{
    /// <summary>Empty string.</summary>
    string string::Empty = make_string("");

    dot::type_t string_impl::typeof()
    {
        static dot::type_t type = []()->dot::type_t
        {
            dot::type_t type = dot::make_type_builder<string_impl>("System", "string")
                ->build();
            return type;
        }();
        return type;
    }

    dot::type_t string_impl::type()
    {
        return typeof();
    }


    bool string_impl::equals(object obj)
    {
        if (this == &(*obj)) return true;

        if (obj.is<string>())
        {
            return *this == *obj.as<string>();
        }

        return false;
    }

    size_t string_impl::hash_code()
    {
        return std::hash<std::string>()(*this);
    }

    string string_impl::to_string()
    {
        return this;
    }

    /// <summary>Determines whether the end of this
    /// string matches the specified string.</summary>
    bool string_impl::ends_with(const string& value)
    {
        int p = length() - value->length();
        if (p >= 0 && substr(p, value->length()) == *value)
            return true;
        return false;
    }

    /// <summary>Determines whether the beginning of this
    /// string matches the specified string.</summary>
    bool string_impl::starts_with(const string& value)
    {
        int p = length() - value->length();
        if (p >= 0 && substr(0, value->length()) == *value)
            return true;
        return false;
    }

    /// <summary>Retrieves a substring which starts at the specified
    /// character position and has the specified length.</summary>
    string string_impl::substring(int startIndex, int length)
    {
        return make_string(this->substr(startIndex, length));
    }

    int string_impl::index_of_any(list<char> anyOf)
    {
        size_t pos = find_first_of(anyOf->data(), 0, anyOf->size());
        if (pos != std::string::npos)
            return pos;
        return -1;
    }

    string string_impl::remove(int startIndex)
    {
        return make_string(*this)->erase(startIndex);
    }

    string string_impl::remove(int startIndex, int count)
    {
        return make_string(*this)->erase(startIndex, count);
    }

    string string_impl::replace(const char oldChar, const char newChar) const
    {
        string new_str = *this;
        std::replace(new_str->begin(), new_str->end(), oldChar, newChar);
        return new_str;
    }

    bool string::is_null_or_empty(string value)
    {
        if (value == nullptr || value->empty())
            return true;
        return false;
    }

    /// <summary>Case sensitive comparison to object.</summary>
    bool string::operator==(const object& rhs) const
    {
        // If rhs is null, return false. Otherwise, check if
        // the other object is a string. If yes, compare by value.
        // If no, return false.
        if (rhs == nullptr)
        {
            return false;
        }
        else
        {
            string rhsStr = rhs.as<string>();
            if (rhsStr != nullptr) return operator==(rhsStr);
            else return false;
        }
    }

    /// <summary>Non-template implementation of string.Format.</summary>
    std::string string::format_impl(fmt::string_view format_str, fmt::format_args args)
    {
        return fmt::vformat(format_str, args);
    }
}
