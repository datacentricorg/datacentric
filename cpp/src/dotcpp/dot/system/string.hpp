/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#pragma once

#include <dot/declare.hpp>
#include <fmt/core.h>
#include <dot/detail/const_string_base.hpp>
#include <dot/system/ptr.hpp>
#include <dot/system/Char.hpp>

namespace dot
{
    class string_impl; class string;
    template <class T> class list_impl; template <class T> using list = ptr<list_impl<T>>;
    template <class T> class nullable;
    enum class string_split_options;
    class Char;

    /// <summary>
    /// Immutable string type.
    ///
    /// The string is encoded internally as UTF-8.
    /// </summary>
    class DOT_CLASS string_impl : public virtual object_impl, public detail::const_string_base
    {
        typedef string_impl self;
        typedef detail::const_string_base base;
        friend string make_string(const std::string& rhs);
        friend string make_string(const char* rhs);

    public: // CONSTRUCTORS

        /// <summary>Creates an empty string.</summary>
        string_impl() {}

        /// <summary>Create from a single Unicode character.</summary>
        string_impl(const Char& value);

        /// <summary>
        /// Initializes from std::string or string literal.
        ///
        /// This constructor is private. Use make_string(str) function instead.
        /// </summary>
        string_impl(const std::string& value) : base(value) {}

        /// <summary>Create from const char*, null pointer is converted to to empty value.</summary>
        string_impl(const char* value) : base(value) {}

        /// <summary>Create from a single 8-bit character.</summary>
        string_impl(char value) : base(std::to_string(value)) {}

    public: // METHODS

        /// <summary>Determines whether this instance and a specified object,
        /// which must also be a string object, have the same value.</summary>
        bool equals(object obj) override;

        /// <summary>Returns the hash code for this string.</summary>
        virtual size_t hash_code() override;

        /// <summary>Returns this instance of string; no actual conversion is performed.</summary>
        virtual string to_string() override;

        /// <summary>Determines whether the end of this
        /// string matches the specified string.</summary>
        bool ends_with(const string& value);

        /// <summary>Determines whether the beginning of this
        /// string matches the specified string.</summary>
        bool starts_with(const string& value);

        /// <summary>
        /// Retrieves a substring which starts at the specified
        /// character position and has the specified length.
        /// </summary>
        string substring(int startIndex, int length);

        /// <summary>
        /// Compares this instance with a specified string object and indicates
        /// whether this instance precedes, follows, or appears in the same position
        /// in the sort order as the specified string.
        /// </summary>
        int compare_to(const string& strB) const; // TODO - implement

        /// <summary>
        /// Reports the zero-based index of the first occurrence in this instance of any
        /// character in a specified array of Unicode characters.
        /// </summary>
        int index_of_any(list<char> anyOf);

        /// <summary>
        /// Returns a new string in which all the characters in the current instance, beginning
        /// at a specified position and continuing through the last position, have been deleted.
        /// </summary>
        string remove(int startIndex);

        /// <summary>
        /// Returns a new string in which a specified number of characters in the current
        /// instance beginning at a specified position have been deleted.
        /// </summary>
        string remove(int startIndex, int count);

        /// <summary>
        /// Returns a new string in which all occurrences of a specified string
        /// in the current instance are replaced with another specified string.
        /// </summary>
        string replace(const string& oldValue, const string& newValue) const; // TODO - implement

        /// <summary>
        /// Returns a new string in which all occurrences of a specified character
        /// in this instance are replaced with another specified character.
        /// </summary>
        string replace(const char oldChar, const char newChar) const;

        /// <summary>
        /// Returns a string array that contains the substrings of the current string
        /// that are delimited by any of the specified 8-bit characters.
        /// </summary>
        list<string> split(char separator) const;  // TODO - implement

        /// <summary>
        /// Returns a string array that contains the substrings in
        /// this string that are delimited by any of the specified strings.
        /// A parameter specifies whether to return empty array elements.
        /// </summary>
        list<string> split(const list<string>& separator, const string_split_options& options) const; // TODO - implement

        ///<summary>Indicates whether the argument occurs within this string.</summary>
        bool contains(string const& s) const; // TODO - implement

        ///<summary>Returns a copy of this string converted to lowercase.</summary>
        string to_lower() const; // TODO - implement

        ///<summary>Returns a copy of this string converted to uppercase.</summary>
        string to_upper() const; // TODO - implement

    public: // OPERATORS

        /// <summary>Gets the Char object at a specified position in the current string object.</summary>
        Char operator[](int index) const;

        /// <summary>Returns a string containing characters from lhs followed by the characters from rhs.</summary>
        friend string operator+(const string& lhs, const string& rhs);

        /// <summary>Returns a string containing characters from lhs followed by the characters from rhs.</summary>
        friend bool operator<(const string& lhs, const string& rhs);

    public: // REFLECTION

        virtual type_t type();
        static type_t typeof();
    };

    /// <summary>
    /// Pointer to string_impl that has additional constructors compared to ptr(string)
    /// </summary>
    class DOT_CLASS string : public ptr<string_impl>
    {
        typedef ptr<string_impl> base;

    public: // CONSTANTS

        /// <summary>Empty string.</summary>
        static string Empty;

    public: // CONSTRUCTORS

        /// <summary>Create null string.</summary>
        string() : base() {}

        /// <summary>
        /// Take ownership of raw pointer to template argument type.
        /// This also permits construction from null pointer.
        /// </summary>
        string(string_impl* p) : base(p) {}

        /// <summary>
        /// Create from std::string.
        /// </summary>
        string(const std::string& rhs) : base(new string_impl(rhs)) {}

        /// <summary>
        /// Create from string literal.
        /// </summary>
        string(const char* rhs) : base(new string_impl(rhs)) {}

        /// <summary>
        /// Copy constructor.
        /// </summary>
        string(const string& rhs) : base(rhs) {}

        /// <summary>
        /// Supports cast (explicit constructor) from object.
        ///
        /// Error if object does is not a boxed int.
        /// Null object becomes empty nullableInt.
        /// </summary>
        explicit string(const ptr<object_impl>& rhs) : base(rhs) {}

    public: // STATIC

        /// <summary>
        /// Replaces format entries in the specified string with the string
        /// representation of objects in the argument array.
        /// </summary>
        template <typename ...Args>
        static string format(const string& format_string, const Args& ...args);

        /// <summary>Indicates whether the specified string is null or an string.Empty string.</summary>
        static bool is_null_or_empty(string value);

    public: // OPERATORS

        /// <summary>Case sensitive comparison to std::string.</summary>
        bool operator==(const std::string& rhs) const { string_impl& impl = base::operator*(); return impl == rhs; }

        /// <summary>Case sensitive comparison to std::string.</summary>
        bool operator!=(const std::string& rhs) const { return !operator==(rhs); }

        /// <summary>Case sensitive comparison to string literal.</summary>
        bool operator==(const char* rhs) const { if (rhs == nullptr) return base::operator==(nullptr); string_impl& impl = base::operator*(); return impl == rhs; }

        /// <summary>Case sensitive comparison to string literal.</summary>
        bool operator!=(const char* rhs) const { return !operator==(rhs); }

        /// <summary>Case sensitive comparison to string literal.</summary>
        bool operator==(const ptr<string_impl>& rhs) const { return *this == *rhs; }

        /// <summary>Case sensitive comparison to string literal.</summary>
        bool operator!=(const ptr<string_impl>& rhs) const { return !operator==(rhs); }

        /// <summary>Case sensitive comparison to object.</summary>
        bool operator==(const object& rhs) const;

        /// <summary>Case sensitive comparison to object.</summary>
        bool operator!=(const object& rhs) const { return !operator==(rhs); }

    private: // STATIC

        /// <summary>Non-template implementation of string.Format.</summary>
        static std::string format_impl(fmt::string_view format_str, fmt::format_args args);
    };

    /// <summary>
    /// Create from std::string or string literal using new
    /// </summary>
    inline string make_string(const std::string& rhs) { return new string_impl(rhs); }

    /// <summary>
    /// Create from std::string or string literal using new
    /// </summary>
    inline string make_string(const char* rhs) { return new string_impl(rhs); }

    /// <summary>Returns a string containing characters from lhs followed by the characters from rhs.</summary>
    inline string operator+(const string& lhs, const string& rhs) { return make_string(*lhs + *rhs); }

    /// <summary>Returns a string containing characters from lhs followed by the characters from rhs.</summary>
    inline bool operator<(const string& lhs, const string& rhs) { return *lhs < *rhs; }

    /// <summary>Helper class for fmt::format arguments conversion</summary>
    template <class T>
    struct format_forward {
        static inline auto convert(const T& t) { return t; }
    };

    /// <summary>Helper class for fmt::format arguments conversion</summary>
    template<>
    struct format_forward<string> {
        static inline const std::string& convert(const string& s) { return *s; }
    };

    /// <summary>Helper class for fmt::format arguments conversion</summary>
    template<class T>
    struct format_forward<ptr<T>> {
        static inline std::string convert(const ptr<T>& o) { return *o->to_string(); }
    };

    /// <summary>Helper class for fmt::format arguments conversion</summary>
    template<class T>
    struct format_forward<nullable<T>> {
        static inline T convert(const nullable<T>& o) { return o.value_or_default(); }
    };

    template<typename ...Args>
    string string::format(const string& format_string, const Args& ...args)
    {
        return format_impl(*format_string, fmt::make_format_args(format_forward<Args>::convert(args)...));
    }
}

namespace std
{
    /// <summary>Implements hash struct used by STL unordered_map for string.</summary>
    template <>
    struct hash<dot::string> : public hash<dot::ptr<dot::string_impl>>
    {};

    /// <summary>Implements equal_to struct used by STL unordered_map for string.</summary>
    template <>
    struct equal_to<dot::string> : public equal_to<dot::ptr<dot::string_impl>>
    {};
}
