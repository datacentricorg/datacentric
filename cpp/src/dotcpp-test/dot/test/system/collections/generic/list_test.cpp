/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <dot/test/implement.hpp>
#include <approvals/ApprovalTests.hpp>
#include <approvals/Catch.hpp>
#include <dot/system/collections/generic/list.hpp>
#include <dot/system/string.hpp>
#include <dot/system/type.hpp>

namespace dot
{
    static std::stringstream received;

    /// <summary>Print double vector to received on one line in JSON format.</summary>
    static void PrintList(string name, list<double> v)
    {
        received << "\"" << *name << "\" : [ ";
        for (int i = 0; i < v->count(); ++i)
        {
            if (i > 0) received << ", ";
            received << v[i];
        }
        received << " ]" << std::endl;
    }

    TEST_CASE("Smoke")
    {
        list<double> a = make_list<double>();
        a->add(0.0);
        a->add(1.0);
        a->add(2.0);

        REQUIRE(a->count() == 3);

    }

    TEST_CASE("Interfaces")
    {
        list<double> a = make_list<double>();

        // Add elements to List interface
        list<double> b = a;
        b->add(2.0);
        b->add(1.0);
        b->add(0.0);

        // Check size of the original class and the interface
        REQUIRE(a->count() == 3);
        REQUIRE(b->count() == 3);

        // Access the underlying std::vector<double> class
        PrintList("Unsorted", a);
        std::sort(a->begin(), a->end());
        PrintList("Sorted", a);

        // Access by object
        object obj = b;
        // TODO - uncomment next line when output matches C#
        // received << obj->to_string() << std::endl;

        Approvals::verify(received.str());
        received.clear();
    }

    TEST_CASE("Iterators")
    {
        list<string> string_list = make_list<string>();
        string_list->add("000");
        string_list->add("111");
        string_list->add("222");
        REQUIRE(string_list->count() == 3);

        int i = 0;
        for (string str : string_list)
        {
            REQUIRE(string_list[i++] == str);
        }
    }

    TEST_CASE("Capacity")
    {
        list<string> string_list = make_list<string>();
        string_list->set_capacity(100);
        REQUIRE(string_list->capacity() == 100);
    }

    TEST_CASE("Find")
    {
        dot::list<string> string_list = make_list<string>();
        string_list->add("000");
        string_list->add("111");
        string_list->add("222");
        string_list->add("333");
        string_list->add("444");

        // TODO string_list.findLast([](std::string const& s) { return s == "222"; }) = "57";
        // TODO REQUIRE(string_list.findLastIndex([](std::string const& s) { return s == "111"; }) == 0);
    }
}
