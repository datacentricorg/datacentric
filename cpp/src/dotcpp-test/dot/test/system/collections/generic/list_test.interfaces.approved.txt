"Unsorted" : [ 2, 1, 0 ]
"Sorted" : [ 0, 1, 2 ]

